app/services/work_package_bulk_update_service.rb

```diff
@@ -0,0 +1,166 @@
+    prepare(params)
+
+    unsaved_work_packages = []
+    moved_work_packages   = []
+    @work_packages.sort!
+
+    @work_packages.each do |work_package|
+      work_package.reload
+
+      if isMoveOrCopy
+        Redmine::Hook.call_hook(:controller_work_packages_move_before_save,
+                                params:         params,
+                                work_package:   work_package,
+                                target_project: @target_project,
+                                copy:           !!@copy)
+
+        moved_wp = work_package.move_to_project(@target_project,
+                                                @target_type,
+                                                copy:         @copy,
+                                                attributes:   permit_params_for_move_or_copy(params),
+                                                journal_note: @notes)
+
+        if moved_wp
+          moved_work_packages << moved_wp
+        else
+          unsaved_work_packages << work_package
+        end
+      else
+        work_package.add_journal(User.current, @notes)
+
+        # filter parameters by whitelist and add defaults
+        attributes = permit_params_for_edit params, work_package.project
+        work_package.assign_attributes attributes
+
+        Redmine::Hook.call_hook(:controller_work_package_bulk_before_save,
+                                params: params, work_package: work_package)
+        JournalObserver.instance.send_notification = params[:send_notification] != '0'
+        unless work_package.save
+          unsaved_work_packages << work_package
+        end
+      end
+    end
+
+    {
+      moved_work_packages:   moved_work_packages,
+      unsaved_work_packages: unsaved_work_packages,
+      copy: @copy
+    }
+  end
+
+  private
+
+  def prepare(params)
+    @work_packages.sort!
+    @copy             = params.has_key? :copy
+    @allowed_projects = WorkPackage.allowed_target_projects_on_move
+    if params[:new_project_id]
+      @target_project = @allowed_projects.detect { |p| p.id.to_s == params[:new_project_id].to_s }
+    end
+
+    # allowed values will be derived from the effective project
+    effective_projects = @target_project ? [@target_project] : @projects
+
+    @custom_fields = intersect_arrays effective_projects.map(&:all_work_package_custom_fields)
+    @assignables   = intersect_arrays effective_projects.map(&:possible_assignees)
+    @responsibles  = intersect_arrays effective_projects.map(&:possible_responsibles)
+    @statuses      = intersect_arrays effective_projects.map { |p| Workflow.available_statuses(p) }
+    @types         = intersect_arrays effective_projects.map(&:types)
+
+    @target_project ||= @project
+    @target_type = params[:new_type_id].nil? ? nil : @types.detect { |t| t.id.to_s == params[:new_type_id].to_s }
+    @notes       = params[:notes] || ''
+
+    self
+  end
+
+  def permit_params_for_edit(params, project)
```

Comment:
Assignment Branch Condition size for permit_params_for_edit is too high. [20.83/15]